File: voyago-frontend/src/services/expenseService.js
import api from './api';

/**
 * Trip Expense API Service
 * Encapsulates backend HTTP calls for Trip Expense management.
 */

/**
 * Create a new expense for a trip.
 * @param {string|number} tripId - Parent Trip ID
 * @param {Object} data - Expense payload { category, amount, description, expenseDate, paymentMethod }
 * @returns {Promise<Object>} Created expense object
 */
export const createExpense = async (tripId, data) => {
  const response = await api.post(`/trips/${tripId}/expenses`, data);
  return response.data;
};

/**
 * Fetch all expenses for a specific trip.
 * @param {string|number} tripId - Parent Trip ID
 * @returns {Promise<Array>} List of expenses ordered by newest date first
 */
export const getExpenses = async (tripId) => {
  const response = await api.get(`/trips/${tripId}/expenses`);
  return response.data;
};

/**
 * Fetch a single expense by ID.
 * @param {string|number} tripId - Parent Trip ID
 * @param {string|number} expenseId - Expense ID
 * @returns {Promise<Object>} Expense details object
 */
export const getExpense = async (tripId, expenseId) => {
  const response = await api.get(`/trips/${tripId}/expenses/${expenseId}`);
  return response.data;
};

/**
 * Update an existing expense.
 * @param {string|number} tripId - Parent Trip ID
 * @param {string|number} expenseId - Expense ID
 * @param {Object} data - Updated expense payload
 * @returns {Promise<Object>} Updated expense object
 */
export const updateExpense = async (tripId, expenseId, data) => {
  const response = await api.put(`/trips/${tripId}/expenses/${expenseId}`, data);
  return response.data;
};

/**
 * Delete an expense.
 * @param {string|number} tripId - Parent Trip ID
 * @param {string|number} expenseId - Expense ID
 * @returns {Promise<void>}
 */
export const deleteExpense = async (tripId, expenseId) => {
  const response = await api.delete(`/trips/${tripId}/expenses/${expenseId}`);
  return response.data;
};

const expenseService = {
  createExpense,
  getExpenses,
  getExpense,
  updateExpense,
  deleteExpense,
};

export default expenseService;

File: voyago-backend/src/main/java/com/voyago/backend/service/ExpenseService.java
package com.voyago.backend.service;

import com.voyago.backend.dto.CreateExpenseRequest;
import com.voyago.backend.dto.ExpenseResponse;
import com.voyago.backend.dto.UpdateExpenseRequest;
import com.voyago.backend.entity.Expense;
import com.voyago.backend.entity.Trip;
import com.voyago.backend.entity.User;
import com.voyago.backend.exception.ResourceNotFoundException;
import com.voyago.backend.repository.ExpenseRepository;
import com.voyago.backend.repository.TripRepository;
import com.voyago.backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ExpenseService {

    private final ExpenseRepository expenseRepository;
    private final TripRepository tripRepository;
    private final UserRepository userRepository;

    @Transactional
    public ExpenseResponse createExpense(Long tripId, CreateExpenseRequest request) {
        Trip trip = getOwnedTrip(tripId);

        Expense expense = Expense.builder()
                .trip(trip)
                .category(request.getCategory().trim().toUpperCase())
                .amount(request.getAmount())
                .description(request.getDescription())
                .expenseDate(request.getExpenseDate())
                .paymentMethod(request.getPaymentMethod())
                .build();

        Expense savedExpense = expenseRepository.save(expense);
        return toResponse(savedExpense);
    }

    @Transactional(readOnly = true)
    public List<ExpenseResponse> getExpenses(Long tripId) {
        getOwnedTrip(tripId);

        List<Expense> expenses = expenseRepository.findByTripIdOrderByExpenseDateDescCreatedAtDesc(tripId);
        return expenses.stream()
                .map(this::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public ExpenseResponse getExpense(Long tripId, Long expenseId) {
        getOwnedTrip(tripId);

        Expense expense = expenseRepository.findByIdAndTripId(expenseId, tripId)
                .orElseThrow(() -> new ResourceNotFoundException("Expense not found with id: " + expenseId));

        return toResponse(expense);
    }

    @Transactional
    public ExpenseResponse updateExpense(Long tripId, Long expenseId, UpdateExpenseRequest request) {
        getOwnedTrip(tripId);

        Expense expense = expenseRepository.findByIdAndTripId(expenseId, tripId)
                .orElseThrow(() -> new ResourceNotFoundException("Expense not found with id: " + expenseId));

        expense.setCategory(request.getCategory().trim().toUpperCase());
        expense.setAmount(request.getAmount());
        expense.setDescription(request.getDescription());
        expense.setExpenseDate(request.getExpenseDate());
        expense.setPaymentMethod(request.getPaymentMethod());

        Expense updatedExpense = expenseRepository.save(expense);
        return toResponse(updatedExpense);
    }

    @Transactional
    public void deleteExpense(Long tripId, Long expenseId) {
        getOwnedTrip(tripId);

        Expense expense = expenseRepository.findByIdAndTripId(expenseId, tripId)
                .orElseThrow(() -> new ResourceNotFoundException("Expense not found with id: " + expenseId));

        expenseRepository.delete(expense);
    }

    private Trip getOwnedTrip(Long tripId) {
        User user = getCurrentUser();
        return tripRepository.findByIdAndUserId(tripId, user.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Trip not found with id: " + tripId));
    }

    private User getCurrentUser() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !authentication.isAuthenticated() || "anonymousUser".equals(authentication.getPrincipal())) {
            throw new IllegalArgumentException("User is not authenticated");
        }

        String email = authentication.getName();
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("User not found with email: " + email));
    }

    private ExpenseResponse toResponse(Expense expense) {
        return ExpenseResponse.builder()
                .id(expense.getId())
                .tripId(expense.getTrip() != null ? expense.getTrip().getId() : null)
                .category(expense.getCategory())
                .amount(expense.getAmount())
                .description(expense.getDescription())
                .expenseDate(expense.getExpenseDate())
                .paymentMethod(expense.getPaymentMethod())
                .createdAt(expense.getCreatedAt())
                .updatedAt(expense.getUpdatedAt())
                .build();
    }
}
